work_queue.cpp

```diff
@@ -0,0 +1,89 @@
+//
+// Created by victor on 6/2/21.
+//
+
+#include "include/work_queue.h"
+
+const unsigned first_idx = 2;
+
+WorkQueue::WorkQueue(uint32_t buffer_size, Node nodes) : buffer_size
+(buffer_size), cq(20,buffer_size*sizeof(Node)), buffers(nodes) {
```

Comment:
Don't hard code the number of slots in the circular queue. This could cause us problems if we run with too many threads. The number of slots should be 2*num_groups